### PR TITLE
internal: add db operations to api

### PR DESCRIPTION
**Pull request by s1na**
See https://github.com/ethereum/go-ethereum/issues/24720.

### CODE REVIEW

**Comment by holiman** [comment]
## Get
The CLI version also accepts non-hex, which is very handy sometimes:
```
[user@work go-ethereum]$ ./build/bin/geth --sepolia db get 0x446174616261736556657273696f6e
INFO [04-22|13:36:48.115] Maximum peer count                       ETH=50 LES=0 total=50
INFO [04-22|13:36:48.115] Smartcard socket not found, disabling    err="stat /run/pcscd/pcscd.comm: no such file or directory"
INFO [04-22|13:36:48.115] Set global gas cap                       cap=50,000,000
INFO [04-22|13:36:48.116] Allocated cache and file handles         database=/home/user/.ethereum/sepolia/geth/chaindata cache=512.00MiB handles=262,144 readonly=true
INFO [04-22|13:36:48.130] Opened ancient database                  database=/home/user/.ethereum/sepolia/geth/chaindata/ancient readonly=true
key 0x446174616261736556657273696f6e: 0x08
[user@work go-ethereum]$ ./build/bin/geth --sepolia db get DatabaseVersion
INFO [04-22|13:36:50.070] Maximum peer count                       ETH=50 LES=0 total=50
INFO [04-22|13:36:50.071] Smartcard socket not found, disabling    err="stat /run/pcscd/pcscd.comm: no such file or directory"
INFO [04-22|13:36:50.074] Set global gas cap                       cap=50,000,000
INFO [04-22|13:36:50.074] Allocated cache and file handles         database=/home/user/.ethereum/sepolia/geth/chaindata cache=512.00MiB handles=262,144 readonly=true
INFO [04-22|13:36:50.080] Opened ancient database                  database=/home/user/.ethereum/sepolia/geth/chaindata/ancient readonly=true
key 0x446174616261736556657273696f6e: 0x08
```
vs
```
> db.get("0x446174616261736556657273696f6e")
"0x08"
> db.get("DatabaseVersion")
WARN [04-22|13:35:40.904] Served db_get                            reqid=10 duration="28.696µs" err="invalid argument 0: json: cannot unmarshal hex string without 0x prefix into Go value of type hexutil.Bytes"
Error: invalid argument 0: json: cannot unmarshal hex string without 0x prefix into Go value of type hexutil.Bytes
	at web3.js:6365:37(47)
	at send (web3.js:5099:62(35))
	at <eval>:1:7(4)

```

## Put
The cli version shows the previous value, which can be useful.
```
[user@work go-ethereum]$ ./build/bin/geth --sepolia db put DatabaseVersion 0x08
INFO [04-22|13:40:08.859] Maximum peer count                       ETH=50 LES=0 total=50
INFO [04-22|13:40:08.859] Smartcard socket not found, disabling    err="stat /run/pcscd/pcscd.comm: no such file or directory"
INFO [04-22|13:40:08.860] Set global gas cap                       cap=50,000,000
INFO [04-22|13:40:08.860] Allocated cache and file handles         database=/home/user/.ethereum/sepolia/geth/chaindata cache=512.00MiB handles=262,144
INFO [04-22|13:40:08.898] Opened ancient database                  database=/home/user/.ethereum/sepolia/geth/chaindata/ancient readonly=false
Previous value: 0x0a
INFO [04-22|13:40:08.899] Freezer shutting down 
```

This one just doesn't return anything, but IMO might be nifty if it returned the previous value. 
```
> db.put('0x446174616261736556657273696f6e','0x0a')
null
```
Not sure if this should be in `db` or if it should be in `debug`. It's definitely a non-standard and dangerous api (not dangerous as in "can lead to RCE", but dangerous as in "can fry your entire database")

**Review by holiman — COMMENTED** [review]
Functionality-wise, this LGTM, with two minor nitpicks. 
Bikeshedding-wise, someone will have to make a call whether it's in the right place or not . Probably @fjl or @karalabe

**Comment by s1na** [comment]
Fixed the comments. Only now `put` returns `0x` if there was no previous value.

File: internal/ethapi/api.go [hunk 1]
```diff
@@ -2021,3 +2066,12 @@ func toHexSlice(b [][]byte) []string {
 	}
 	return r
 }
+
+// ParseHexOrString tries to hexdecode b, but if the prefix is missing, it instead just returns the raw bytes
+func parseHexOrString(str string) ([]byte, error) {
```
Comment (rjl493456442) [inline internal/ethapi/api.go:2071]: Nitpick, perhaps we can move this function into common package, but feel free to ignore this comment.

File: internal/ethapi/backend.go [hunk 2]
```diff
@@ -137,6 +137,11 @@ func GetAPIs(apiBackend Backend) []rpc.API {
 			Version:   "1.0",
 			Service:   NewPrivateAccountAPI(apiBackend, nonceLock),
 			Public:    false,
+		}, {
+			Namespace: "db",
+			Version:   "1.0",
+			Service:   NewPublicDbAPI(apiBackend),
+			Public:    true,
```
Comment (rjl493456442) [inline internal/ethapi/backend.go:144]: Perhaps we should convert it to private API? Private APIs can also be exposed(http, ws) IIUC if users specify it, Public APIs will be exposed if the API module list specified by users is empty.

File: internal/ethapi/api.go [hunk 3]
```diff
@@ -1998,6 +1998,51 @@ func (s *PublicNetAPI) Version() string {
...
+	}
+	return ret, db.Put(blob, value)
+}
+
+// Delete removes an item from the database.
+func (api *PublicDbAPI) Delete(key string) error {
+	blob, err := parseHexOrString(key)
+	if err != nil {
+		return err
+	}
```
Comment (rjl493456442) [inline internal/ethapi/api.go:2042]: Can we print some logs for database mutation? (for example, the key of deleted item). I think it's fine to dump out logs for this special debug purpose API.

cc @holiman

File: internal/ethapi/api.go [hunk 4]
```diff
@@ -1998,6 +1998,51 @@ func (s *PublicNetAPI) Version() string {
...
+	)
+	blob, err := parseHexOrString(key)
+	if err != nil {
+		return nil, err
+	}
+	data, err := db.Get(blob)
+	if err == nil {
+		ret = data
+	}
+	return ret, db.Put(blob, value)
```
Comment (rjl493456442) [inline internal/ethapi/api.go:2034]: Can we print some logs for database mutation? (for example, key value of the inserted item). I think it's fine to dump out logs for this special debug purpose API.

File: internal/ethapi/api.go [hunk 5]
```diff
@@ -1998,6 +1998,51 @@ func (s *PublicNetAPI) Version() string {
...
+}
+
+// Delete removes an item from the database.
+func (api *PublicDbAPI) Delete(key string) error {
```
Comment (rjl493456442) [inline internal/ethapi/api.go:2038]: Should we return the previous value as well? So that users can see clearly what's deleted.

**Comment by fjl** [comment]
We agreed to

- only expose `get`
- move this functionality to the `debug` namespace, so the name should be `debug_dbGet`.
-

**Review by holiman — APPROVED** [review]
LGTM

File: internal/ethapi/api.go [hunk 6]
```diff
@@ -1972,6 +1972,15 @@ func (api *PrivateDebugAPI) SetHead(number hexutil.Uint64) {
 	api.b.SetHead(uint64(number))
 }
 
+// Get returns the raw value of a key stored in the database.
```
Comment (holiman) [inline internal/ethapi/api.go:1975]:
```suggestion
// DbGet returns the raw value of a key stored in the database.
```

**Review by holiman — COMMENTED** [review]
One little nit

**Pull request merged**